parcels/collection/collectionaos.py

```diff
@@ -484,6 +533,206 @@ def add_same(self, same_class):
+                mpi_comm.Bcast(ids, root=0)
+                pu_ids = ids
+                new_lastID = 0
+                if mpi_rank == 0:
+                    new_lastID = ScipyParticle.lastID+data_array['lon'].shape[0]-1
+                new_lastID = mpi_comm.bcast(new_lastID, root=0)
+                self._pclass.setLastID(new_lastID)
+                pu_center = np.array(np.mean(spdata, axis=0), dtype=self._lonlatdepth_dtype)
+                n_pu_data = pu_indices.shape[0]
+                pu_ncenters = None
+                if mpi_rank == 0:
+                    pu_ncenters = np.empty([mpi_size, pu_center.shape[0]], dtype=self._latlondepth_dtype)
+                mpi_comm.Gather(pu_center, pu_ncenters, root=0)
+                pu_ndata = mpi_comm.gather(n_pu_data, root=0)
+                if mpi_rank == 0:
+                    for i in range(self._pu_centers.shape[0]):
+                        ax = float(pu_ndata[i]) / float(len(np.nonzero(self._pu_indicators == i)[0]))
+                        self._pu_centers[i, :] += ax*pu_ncenters[i, :]
+                mpi_comm.Bcast(self._pu_centers, root=0)
+            else:
+                pu_ids = np.arange(ScipyParticle.lastID, stop=ScipyParticle.lastID+data_array['lon'].shape[0]) if 'id' not in data_array.keys() else data_array['id']
+                new_lastID = ScipyParticle.lastID+data_array['lon'].shape[0]-1
+                self._pclass.setLastID(new_lastID)
+                pu_indices = np.arange(start=0, stop=data_array['lon'].shape[0])
+                n_pu_data = pu_indices.shape[0]
+            if n_pu_data <= 0:
+                results = []
+            else:
+                for i in pu_indices:
+                    pid = pu_ids[i]
+                    pdata = self._pclass(lon=data_array['lon'][i], lat=data_array['lat'][i], pid=pid, ngrids=self._ngrid)
+                    if 'depth' in data_array.keys():
+                        pdata.depth = data_array['depth'][i]
+                    if 'time' in data_array.keys():
+                        pdata.time = data_array['time'][i]
+                    if 'dt' in data_array.keys():
+                        pdata.dt = data_array['dt'][i]
+                    for key in self._kwarg_keys:
+                        if key in data_array.keys():
+                            setattr(pdata, key, data_array[key][i])
+                    results.append(self.add_single(pdata, pu_checked=True))
+        self._ncount = len(self._data)
+        return results
+
+    def add_single(self, particle_obj, pu_checked=False):
+        """
+        Adding a single Particle to the collection - either as a 'Particle; object in parcels itself, or
+        via its ParticleAccessor.
+        :returns index of added particle
+        """
+        # ==== first approach - still need to incorporate the MPI re-centering ==== #
+        super().add_single(particle_obj)
+        assert isinstance(particle_obj, ScipyParticle)
```

Comment:
it is indeed the latter - the `JITParticle` is a derived class of the `ScipyParticle`, which itself is derived of the `Particle` (or: `_Particle` - unsure right now which one it is). 

The `Particle` only includes the barebones of a particle - i.e. the fact that a `Particle` contains a list of `variables`, which themselves are NumPy-mapped instances of the `Variable` class. Hence follows that a `Particle` does not contain any specific variables, also not the minimal set we require for a working `Particle`. Strictly speaking: `Particle` the way it is written now would ideally be tagged as `@abstractclass`, because it cannot be workably instanciated. The first working, non-abstract subclass of a `Particle` is the `ScipyParticle`, which is why we are checking for that class specifically here. In conclusion, the `isinstance` call returns `True` if the object in question is of the type `ScipyParticle` or one if its derivatives (e.g. `JITParticle` as well as the user-customized particle classes).

I hope that explains the thought process behind the assert. Again, I think the one thing one could ideally do to clarify this further with minimal effort is to declare the `Particle` class as abstract.

Reply:
Yes, of course I undersatand the thought process behind the assert. My point was that perhaps we should make `JITParticle` a derivative of `Particle` (which would then be a non-abstract subclass of `_Particle`), so that `ScipyParticle` and `JITParticle` are on the same level in terms of inheritance. 
Not a deal-breaker for now; more something to consider for a next PR?